dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp

```diff
@@ -1546,31 +1515,55 @@ generateF90InterfaceSI(FortranInterfaceModuleGen& fimGen,
           ,
           stencilInstantiation->getMetaData().getFieldDimensions(fieldID).rank());
     }
+    for(auto fieldID : stencilInstantiation->getMetaData().getAPIFields()) {
+      if(includeOuts && (stencils[0]->getFields().at(fieldID).field.getIntend() ==
+                              dawn::iir::Field::IntendKind::InputOutput || stencils[0]->getFields().at(fieldID).field.getIntend() ==
+                              dawn::iir::Field::IntendKind::Output)) {
+        api.addArg(
+            stencilInstantiation->getMetaData().getNameFromAccessID(fieldID) + "_before",
+            FortranAPI::InterfaceType::DOUBLE /* Unfortunately we need to know at codegen
+                                                          time whether we have fields in SP/DP */
+            ,
+            stencilInstantiation->getMetaData().getFieldDimensions(fieldID).rank());
+      }
+    }
   };
 
-  for(auto&& interface : interfaces) {
-    addArgsToAPI(interface);
-    fimGen.addInterfaceAPI(std::move(interface));
-  }
+  addArgsToAPI(interfaces[fromDeviceAPIIdx], /*includeOuts*/ false);
+  fimGen.addInterfaceAPI(std::move(interfaces[fromDeviceAPIIdx]));
+  addArgsToAPI(interfaces[fromHostAPIIdx], /*includeOuts*/ false);
+  fimGen.addInterfaceAPI(std::move(interfaces[fromHostAPIIdx]));
+  addArgsToAPI(interfaces[runAndVerifyIdx], /*includeOuts*/ true);
+  fimGen.addInterfaceAPI(std::move(interfaces[runAndVerifyIdx]));
 
-  addArgsToAPI(runWrapper);
+  addArgsToAPI(runWrapper, /*includeOuts*/ true);
 
-  auto getFieldArgs = [&]() -> std::vector<std::string> {
+  auto getFieldArgs = [&](bool includeOuts) -> std::vector<std::string> {
     std::vector<std::string> args;
     for(auto fieldID : stencilInstantiation->getMetaData().getAPIFields()) {
       args.push_back(stencilInstantiation->getMetaData().getNameFromAccessID(fieldID));
     }
+    for(auto fieldID : stencilInstantiation->getMetaData().getAPIFields()) {
+      if(includeOuts && (stencils[0]->getFields().at(fieldID).field.getIntend() ==
+                              dawn::iir::Field::IntendKind::InputOutput || stencils[0]->getFields().at(fieldID).field.getIntend() ==
+                              dawn::iir::Field::IntendKind::Output)) {
+        args.push_back(stencilInstantiation->getMetaData().getNameFromAccessID(fieldID) +
+                       "_before");
+      }
+    }
     return args;
   };
 
-  auto genCallArgs = [&](FortranWrapperAPI& wrapper, std::string first = "") {
+  auto genCallArgs = [&](FortranWrapperAPI& wrapper, std::string first = "",
+                         bool includeInouts = false) {
```

Comment:
do you need it? you call it always with `includeInouts=true`

Reply:
I do need it, c.f. line 1551 `genCallArgs(runWrapper);` . However, I decided to remove the default argument and make this explicit.